### ISSUE TITLE
Code unreachable according to mypyc but not according to mypy

### ISSUE THREAD

**Issue by twoertwein** [python/mypy#10703]
**Bug Report**

Different behavior between mypy and mypyc.

**To Reproduce**

```py
from pathlib import Path
from typing import TypeVar

PathNone = TypeVar("PathNone", Path, None)


def test(path: PathNone) -> PathNone:
    if path is None:
        return path  # is reachable
    return path / "test"
```
```sh
$ mypy test.py
Success: no issues found in 1 source file
$ mypyc test.py
test.py:9: error: Statement is unreachabl
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: -
- Mypy configuration options from `pyproject.toml`: `warn_unreachable = true`
- Python version used: 3.9.4
- Operating system and version: 4.13.0-36-generic

**Comment by twoertwein** [comment]
Works with 0.931
